### PR TITLE
Referencables verifer

### PR DESCRIPTION
**Pull request by sirherrbatka**
There is no tabs here. If there is odd formatting, could you give me the concrete line number?

### CODE REVIEW

**Comment by zinnschlag** [comment]
I don't see any tab problems in this version. However what I see are random changes in the source formatting (inserting/deleting space and such) of otherwise unchanged code, which make it pretty much impossible to review your code (have a look at the diff; a total mess). Not to mention that these formatting changes are not covered by our guidelines, which means the next person who works on the code may just change them back.

It is okay to make beautifying adjustments to the lines you are working on anyway, but what I see here is a no go.

I did manage to find one more actual problem with the code: You are manually going through the collections to check if an ID exists. We have the searchId function for that, which will also take care of case issues.

Also, the code is very repetitive, which is definitely not a good thing (some templates might come in handy).

**Comment by sirherrbatka** [comment]
Corrected formatting, replaced manual loop with searchId.

I guess that this problem is a potential use case for templates, however ─ I 
didn't learned templates (yet). It will take quite a lot of time for me to 
finish with learning git-fu, dealing with some personal issues and then move to 
templates. If you don't mind waiting for those, just close pull request and be 
patient ;-)

**Comment by zinnschlag** [comment]
Sure. No hurry. But we can leave the pull request open. Its not like it takes up space or anything.

**Comment by sirherrbatka** [comment]
Ok, It's nice to have a real, production code to train new skill, and code 
review at the same time ;-)

File: apps/opencs/model/tools/referenceablecheck.cpp [hunk 1]
```diff
@@ -0,0 +1,984 @@
...
+    CSMWorld::UniversalId id(CSMWorld::UniversalId::Type_Npc, NPC.mId);
+
+    short level(NPC.mNpdt52.mLevel);
+    char Disposition(NPC.mNpdt52.mDisposition);
+    char Reputation(NPC.mNpdt52.mReputation);
+    char Rank(NPC.mNpdt52.mRank);
+    //Don't know what unknown is for
+    int Gold(NPC.mNpdt52.mGold);
+
+    if (NPC.mNpdtType == 12) //12 = autocalculated
```
Comment (lgromanowski) [inline apps/opencs/model/tools/referenceablecheck.cpp]: Use enum instead of plain "12"

File: apps/opencs/model/tools/referenceablecheck.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include "../doc/stage.hpp"
+#include "../world/data.hpp"
+#include "../world/refiddata.hpp"
+
+namespace CSMTools
+{
+    class ReferenceableCheckStage : public CSMDoc::Stage
+    {
+        public:
+            ReferenceableCheckStage(const CSMWorld::RefIdData& referenceable, const CSMWorld::IdCollection<ESM::Race>& races, const CSMWorld::IdCollection<ESM::Class>& classes, const CSMWorld::IdCollection<ESM::Faction>& factions);
```
Comment (lgromanowski) [inline apps/opencs/model/tools/referenceablecheck.hpp]: Please break long lines near 80-100 column.

Comment (zinnschlag) [inline apps/opencs/model/tools/referenceablecheck.hpp]: More 100 than 80. 80 as a character per line limit is rather archaic.

File: apps/opencs/model/tools/referenceablecheck.cpp [hunk 3]
```diff
@@ -0,0 +1,1107 @@
+#include "referenceablecheck.hpp"
+#include "../world/record.hpp"
+#include "../world/universalid.hpp"
+#include <boost/algorithm/string.hpp>
+#include "../../../../openmw/apps/openmw/mwclass/misc.hpp"
```
Comment (zinnschlag) [inline apps/opencs/model/tools/referenceablecheck.cpp]: This line is wrong. First it doesn't compile. Second, you are not supposed to cross-include between different applications in the apps directory.

Comment (sirherrbatka) [inline apps/opencs/model/tools/referenceablecheck.cpp]: I blame scrawl. ;-)

Comment (ghost) [inline apps/opencs/model/tools/referenceablecheck.cpp]: :D 
the include you wanted to use was components/misc/stringops.hpp I guess. But it looks like it's getting implicitely included anyways.

**Comment by zinnschlag** [comment]
Not yet there.
1. I am getting tons of errors about NPCs having an invalid race. Looks like a bug.
2. I think lights with 0 duration are fine (burned out light sources?).

**Comment by sirherrbatka** [comment]
@1 Can't replicate this on my branch.
@2 Burned out lights makes no sense in the morrowind since after getting 
burned out they are removed from the inventory. Using 0 duration light would 
obviously instantly remove it from the inventory ─ very surprising and 
unexpected behavior. Therefore lights should have either negative, or positive 
duration.

**Comment by ghost** [comment]
Hm, uesp says:
"A value of 0 indicates lights which are already burned out."
And lists light_com_torch_burnedout_01 with a time of 0.
Perhaps our implementation is wrong? Should burned out torches maybe not be removed from the inventory? Did anyone test that?

**Comment by sirherrbatka** [comment]
This is vanilla behavior. There are only 3 burned out lights in the 
morrowind.esm and I'm not even sure if they are actually placed in game 
itself.

**Comment by ghost** [comment]
Apparently if you add light_com_torch_burnedout_01 to your inventory and equip it, vanilla just crashes.
Guess you are right, then.

**Comment by sirherrbatka** [comment]
Yeah, Morrowind.esm has a bit of broken stuff inside of it.

Anyway, @zinnschlag do you have those errors after running verifer on morrowind.esm or any other esm file?

**Comment by zinnschlag** [comment]
Morrowind.esm. My copy of Morrowind.esm. Different versions/localisations of that file causing problems again, maybe?

**Comment by sirherrbatka** [comment]
Dnia wtorek, 14 stycznia 2014 06:37:25 Marc Zinnschlag pisze:

> Morrowind.esm. My copy of Morrowind.esm. Different versions/localisations of
> that file causing problems again, maybe?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/zinnschlag/openmw/pull/1195#issuecomment-32269137

Yes, I see. Extreamly stupid mistake on my side. Solved now.

**Pull request merged**